File: benchmark_matmul.py
#!/usr/bin/env python3
"""
Benchmark script for large matrix multiplication and eigendecomposition.

This script benchmarks:
1. Matrix multiplication with a large matrix X (100000x1000)
2. Eigendecomposition of C = X'*X (1000x1000)

Uses pyperf for robust benchmarking with proper warmup and statistics.
"""

import numpy as np
import pyperf


# Global variables to avoid recreation overhead in benchmark loops
X = None
C = None


def setup_matrices():
    """Setup matrices once before benchmarking."""
    global X, C
    n_rows = 100000
    n_cols = 1000
    
    X = np.random.randn(n_rows, n_cols)
    
    C = X.T @ X


def benchmark_matmul(loops):
    """Benchmark function for matrix multiplication."""
    t0 = pyperf.perf_counter()
    for _ in range(loops):
        X.T @ X
    return pyperf.perf_counter() - t0


def benchmark_eigendecomposition(loops):
    """Benchmark function for eigendecomposition."""
    t0 = pyperf.perf_counter()
    for _ in range(loops):
        np.linalg.eigh(C)
    return pyperf.perf_counter() - t0


def main():
    """Main benchmarking routine using pyperf."""
    print("=" * 70)
    print("Matrix Multiplication and Eigendecomposition Benchmark (pyperf)")
    print("=" * 70)
    print()
    
    # Setup matrices once
    setup_matrices()
    
    # Create runner
    runner = pyperf.Runner()
    
    # Benchmark 1: Matrix multiplication X'*X
    print("-" * 70)
    print("BENCHMARK 1: Matrix Multiplication (X'*X)")
    print(f"Matrix X shape: {X.shape}")
    print("-" * 70)
    runner.bench_time_func('matmul', benchmark_matmul)
    
    # Benchmark 2: Eigendecomposition of C
    print()
    print("-" * 70)
    print("BENCHMARK 2: Eigendecomposition of C")
    print(f"Matrix C shape: {C.shape}")
    print("-" * 70)
    runner.bench_time_func('eigendecomposition', benchmark_eigendecomposition)


if __name__ == "__main__":
    main()

File: benchmark_matmul.jl
#!/usr/bin/env julia
"""
Benchmark script for large matrix multiplication and eigendecomposition.

This script benchmarks:
1. Matrix multiplication with a large matrix X (100000x1000)
2. Eigendecomposition of C = X'*X (1000x1000)

Uses BenchmarkTools for robust benchmarking with proper warmup and statistics.
"""

using LinearAlgebra
using BenchmarkTools
using Printf


# Global variables to store matrices
X = Matrix{Float64}(undef, 0, 0)
C = Matrix{Float64}(undef, 0, 0)


"""
    setup_matrices!()

Setup matrices once before benchmarking.
"""
function setup_matrices!(X, C)
    global X, C
    n_rows = 100000
    n_cols = 1000
    
    X .= randn(n_rows, n_cols)
    
    mul!(C, X', X)
end


"""
    benchmark_matmul()

Benchmark function for matrix multiplication.
"""
function benchmark_matmul()
    return X' * X
end


"""
    benchmark_eigendecomposition()

Benchmark function for eigendecomposition.
"""
function benchmark_eigendecomposition()
    return eigen(Symmetric(C))
end


"""
    main()

Main benchmarking routine using BenchmarkTools.
"""
function main()
    println("="^70)
    println("Matrix Multiplication and Eigendecomposition Benchmark (BenchmarkTools)")
    println("="^70)
    println()
    
    # Setup matrices once
    setup_matrices!()
    
    # Benchmark 1: Matrix multiplication X'*X
    println("-"^70)
    println("BENCHMARK 1: Matrix Multiplication (X'*X)")
    println("Matrix X shape: $(size(X))")
    println("-"^70)
    println("Running benchmark with warmup and multiple samples...")
    matmul_bench = @benchmark benchmark_matmul()
    println("\nMatrix Multiplication Results:")
    display(matmul_bench)
    println("\n")
    
    # Benchmark 2: Eigendecomposition of C
    println()
    println("-"^70)
    println("BENCHMARK 2: Eigendecomposition of C")
    println("Matrix C shape: $(size(C))")
    println("-"^70)
    println("Running benchmark with warmup and multiple samples...")
    eigen_bench = @benchmark benchmark_eigendecomposition()
    println("\nEigendecomposition Results:")
    display(eigen_bench)
    println("\n")
    
    # Summary
    println()
    println("="^70)
    println("SUMMARY")
    println("="^70)
    @printf("Matrix multiplication median time:  %.4f seconds\n", median(matmul_bench).time / 1e9)
    @printf("Matrix multiplication mean time:    %.4f seconds\n", mean(matmul_bench).time / 1e9)
    @printf("Eigendecomposition median time:     %.4f seconds\n", median(eigen_bench).time / 1e9)
    @printf("Eigendecomposition mean time:       %.4f seconds\n", mean(eigen_bench).time / 1e9)
    println("="^70)
end

main()